python/bap.py

```diff
@@ -0,0 +1,342 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import os, time, atexit
+import requests
+from subprocess import Popen
+from mmap import mmap
+from urlparse import urlparse, parse_qs
+from tempfile import NamedTemporaryFile
+import json
+import adt, arm, asm, bil
+
+
+__all__ = ["disasm", "image"]
+
+DEBUG_LEVEL = ["Critical", "Error"]
+
+instance = None
+
+def del_instance():
+    if instance is not None:
+        instance.close()
+
+def get_instance(**kwargs):
+    global instance
+    if 'server' in kwargs or instance is None:
+        if instance is not None:
+            instance.close()
+        args = kwargs.get('server', {})
+        instance = Bap(args)
+    return instance
+
+atexit.register(del_instance)
+
+def disasm(obj, **kwargs):
+    r""" disasm(obj) disassembles provided object.
+    Returns a generator object yield instructions.
+
+    """
+    def ret(obj):
+        return get_instance(**kwargs).insns(obj)
+    if isinstance(obj, Id):
+        return ret(obj)
+    elif isinstance(obj, Resource):
+        return ret(obj.ident)
+    else:
+        return ret(load_chunk(obj, **kwargs))
+
+def image(f, **kwargs):
+    bap = get_instance(**kwargs)
+    if os.path.isfile(f) and not os.path.isabs(f):
+        f = os.path.abspath(f)
+    return Image(bap.load_file(f), bap)
+
+def load_chunk(s, **kwargs):
+    return get_instance(**kwargs).load_chunk(s, **kwargs)
+
+
+class Resource(object):
+    def __init__(self, name, ident, bap):
+        self.ident = Id(ident)
+        self.bap = bap
+        self.msg = None
+        self._name = name
+
+    def load(self):
+        if self.msg is None:
+            self.msg = self.bap.get_resource(self.ident)
+            if not self._name in self.msg:
+                if 'error' in msg:
+                    raise ServerError(response)
+                else:
+                    msg = "Expected {0} msg but got {1}".format(
+                        self._name, msg)
+                    raise RuntimeError(msg)
+
+    def get(self, child):
+        self.load()
+        return self.msg[self._name].get(child)
+
+
+class Image(Resource):
```

Comment:
Does Image/Section/Symbol really go in here?
I think it probably goes in a nested module. That module might want to be pulled up to the base namespace, but code for "Here is how we will connect to the bap server" should probably be separated from "Here are a number of translations for specific queries we want to make"